### PR TITLE
JDK-8330565 : C2: Multiple crashes with CTW after JDK-8316991

### PR DESCRIPTION
**Pull request by JohnTortugo**
The `# assert(false) failed: Bad graph detected in build_loop_late` failure was caused because a string concatenation optimization using [this method](https://github.com/openjdk/jdk/blob/819f3d6fc70ff6fe54ac5f9033c17c3dd4326aa5/src/hotspot/share/opto/graphKit.cpp#L4115) adds AddP and LoadN nodes to IR graph as NotNull _and_ because RAM was not "nullyfing" phis merging nullable pointers. I was only able to reproduce this problem using a classfile/jar compiled using an "old" version of JDK.. because newer version use InvokeDynamic to do string concatenation.

Tested with JTREG tier1-4 on Linux x86_64 & ARM64.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330565](https://bugs.openjdk.org/browse/JDK-8330565): C2: Multiple crashes with CTW after JDK-8316991 (**Bug** - P2)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [94eb0e12](https://git.openjdk.org/jdk/pull/19147/files/94eb0e122e19014783b5a2bf8805842eb3c95e97)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19147/head:pull/19147` \
`$ git checkout pull/19147`

Update a local copy of the PR: \
`$ git checkout pull/19147` \
`$ git pull https://git.openjdk.org/jdk.git pull/19147/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19147`

View PR using the GUI difftool: \
`$ git pr show -t 19147`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19147.diff">https://git.openjdk.org/jdk/pull/19147.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19147#issuecomment-2101675298)

### CODE REVIEW

File: src/hotspot/share/opto/escape.cpp [hunk 1]
```diff
@@ -763,6 +763,21 @@ Node* ConnectionGraph::split_castpp_load_through_phi(Node* curr_addp, Node* curr
...
+    if (load_type->isa_narrowoop()) {
+      new_t = load_type->meet(TypeNarrowOop::NULL_PTR);
+    } else if (load_type->isa_narrowklass()) {
+      new_t = load_type->meet(TypeNarrowKlass::NULL_PTR);
+    } else if (load_type->isa_ptr()) {
+      new_t = load_type->meet(TypePtr::NULL_PTR);
+    }
+    _igvn->set_type(data_phi,  new_t);
+    data_phi->raise_bottom_type(new_t);
+  }
```
Comment (vnkozlov) [inline src/hotspot/share/opto/escape.cpp:779]: Do you intentionally execute `_igvn->transform(` for `data_phi` before you set inputs and now type?
Usually we do transform after we fully construct node.

Comment (JohnTortugo) [reply src/hotspot/share/opto/escape.cpp]: I used to call `transform` right after the nodes were created. Thanks for clarifying on the right way to use it. I refactored this method to call `transform` only after the nodes have all inputs set - doing that alone doesn't fix the problem, though.

**Review by vnkozlov — COMMENTED** [review]
Looks good. This have to be tested.

File: test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndNullableLoads.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+/*
+ * @test
+ * @bug 8330565
+ * @summary Check that Reduce Allocation Merges does not crash when legacy
+ *          string concatenation optimization is applied.
+ * @requires vm.flagless & vm.compiler2.enabled & vm.opt.final.EliminateAllocations
+ * @compile -XDstringConcat=inline TestReduceAllocationAndNullableLoads.java
+ * @run main/othervm -XX:CompileCommand=compileonly,*TestReduceAllocationAndNullableLoads*::*
+ *                   -XX:CompileCommand=dontinline,*TestReduceAllocationAndNullableLoads*::*
+ *                   -XX:-TieredCompilation -Xbatch -Xcomp -server
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndNullableLoads.java:33]:
```suggestion
 *                   -XX:-TieredCompilation -Xcomp -server
```

`-Xcomp` implies `-Xbatch`

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too. I submitted testing.

**Comment by JohnTortugo** [comment]
@TobiHartmann, did the tests pass? Thanks.

**Comment by TobiHartmann** [comment]
Yes, all tests passed. Sorry for the delay. @vnkozlov needs to press approve as well.

**Review by vnkozlov — APPROVED** [review]
Good

**Comment by JohnTortugo** [comment]
Thank you all for reviewing!

**Pull request closed without merging**